src/jsoncanvas.rs

```diff
@@ -2,34 +2,46 @@ use std::collections::HashMap;
 use std::fmt::{Display, Formatter};
 use std::str::FromStr;
 
+use crate::edge::Edge;
 use crate::node::GenericNodeInfo;
 use crate::node::Node;
-use crate::edge::Edge;
-use crate::NodeId;
 use crate::EdgeId;
+use crate::NodeId;
 
-use serde::{Serialize, Serializer, Deserialize, Deserializer};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde_json;
+use thiserror::Error;
 
-#[derive(Debug)]
+#[derive(Debug, Error)]
 pub enum JsonCanvasError {
-    AlreadyExists,
-    NodeNotExists,
-    ParseError
+    #[error("Node {0} already exists")]
+    NodeExists(NodeId),
+    #[error("Edge {0} already exists")]
+    EdgeExists(EdgeId),
+    #[error("Node {0} does not exist")]
+    NodeNotExists(NodeId),
+    #[error(transparent)]
+    ParseError(#[from] serde_json::Error),
 }
 
-
 /// JsonCanvas
 ///
 /// Main struct for the canvas
 ///
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize, Default)]
```

Comment:
Is that a common Rust recommendation to prefer Default instead of new?

Reply:
Yes, it's generally preferred. There's a trait for it which plays very nicely with a lot of other generic functions and combinators - [std::mem::take](https://doc.rust-lang.org/std/mem/fn.take.html), [Entry::or_default](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_default). While a lot of std types have a zero-arg `::new()` which does the same thing, my understanding is that those predate the design of `Default` and are soft-deprecated.